geode-core/src/integrationTest/java/org/apache/geode/management/internal/FederatingManagerConcurrencyIntegrationTest.java

```diff
@@ -0,0 +1,111 @@
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.management.internal;
+
+import static org.apache.geode.distributed.ConfigurationProperties.LOCATORS;
+import static org.apache.geode.internal.net.SocketCreator.getLocalHost;
+import static org.apache.geode.management.internal.SystemManagementService.FEDERATING_MANAGER_FACTORY_PROPERTY;
+import static org.apache.geode.test.awaitility.GeodeAwaitility.await;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.net.UnknownHostException;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.contrib.java.lang.system.RestoreSystemProperties;
+import org.junit.experimental.categories.Category;
+
+import org.apache.geode.StatisticsFactory;
+import org.apache.geode.cache.CacheFactory;
+import org.apache.geode.distributed.internal.InternalDistributedSystem;
+import org.apache.geode.distributed.internal.membership.InternalDistributedMember;
+import org.apache.geode.internal.cache.InternalCache;
+import org.apache.geode.internal.statistics.StatisticsClock;
+import org.apache.geode.management.ManagementService;
+import org.apache.geode.test.junit.categories.JMXTest;
+
+@Category(JMXTest.class)
+public class FederatingManagerConcurrencyIntegrationTest {
+
+  private InternalCache cache;
+  private FederatingManager federatingManager;
+
+  @Rule
+  public RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties();
+
+  @Before
+  public void setUp() {
+    System.setProperty(FEDERATING_MANAGER_FACTORY_PROPERTY,
+        FederatingManagerFactoryWithMockMessenger.class.getName());
+
+    cache = (InternalCache) new CacheFactory()
+        .set(LOCATORS, "")
+        .create();
+
+    SystemManagementService managementService =
+        (SystemManagementService) ManagementService.getExistingManagementService(cache);
+    managementService.createManager();
+    federatingManager = managementService.getFederatingManager();
+    federatingManager.startManager();
+  }
+
+  @After
+  public void tearDown() {
+    cache.close();
+  }
+
+  @Test
+  public void testFederatingManagerConcurrency() throws UnknownHostException {
```

Comment:
Sorry, I am unable to figure out how this is doing operations concurrently, would you be able to explain?  Also, what are we asserting, that the manager can add all the members without exception?

Reply:
@jhuynh1 This was an existing test... We didn't modify this class except for renaming it and changing the method signature in `FederatingManagerFactoryWithMockMessenger`. If you look at just the first commit, you will be able to tell. Sorry for the confusion — that's just how git handles renames and is why we did the first commit separately.

Reply:
The concurrency comes into play because each invocation of `FederatingManager.addMember()` schedules a GII task to be run asynchronously on a thread pool. This test is scheduling a hundred of them.